### PR TITLE
[KEYCLOAK-4900] Pass login_hint parameter to idp

### PR DESCRIPTION
**Pull request by dmnboutin**
Hi, job done to pass login to idp.
Let me know if i have to do some modification.
Regards

### CODE REVIEW

**Comment by stianst** [comment]
A better way would be to just add the login_hint in AbstractOAuth2IdentityProvider.createAuthorizationUrl like so:
.queryParam(OIDCLoginProtocol.LOGIN_HINT_PARAM, request.getAuthenticationSession().getClientNote(OIDCLoginProtocol.LOGIN_HINT_PARAM))

That's a one line change rather than passing it around.

We would also need a test added for this.

**Comment by dmnboutin** [comment]
Hello,
ok so no toggle in the identity provider to pass this value or not ?

**Comment by stianst** [comment]
Yes, please add the toggle, but you don't need to pass the query param around as it'll be already set in the authentication session.

**Comment by stianst** [comment]
Ping - will you update this to use the note from the authentication session and add tests?

**Comment by dmnboutin** [comment]
Hi, 
Yes I will push changes as soon as possible.

**Comment by dmnboutin** [comment]
Hi,
I have pushed modification

File: services/src/main/java/org/keycloak/forms/login/freemarker/FreeMarkerLoginFormsProvider.java [hunk 1]
```diff
@@ -143,6 +144,7 @@ private Response createResponse(LoginFormsPages page) {
         RealmModel realm = session.getContext().getRealm();
         ClientModel client = session.getContext().getClient();
         UriInfo uriInfo = session.getContext().getUri();
+        String loginHint = uriInfo.getQueryParameters().getFirst(OIDCLoginProtocol.LOGIN_HINT_PARAM);
```
Comment (stianst) [inline services/src/main/java/org/keycloak/forms/login/freemarker/FreeMarkerLoginFormsProvider.java]: This shouldn't be needed. In fact you can't rely on the query param being set at this point due to the way authentication flows work.

File: services/src/main/java/org/keycloak/forms/login/freemarker/FreeMarkerLoginFormsProvider.java [hunk 2]
```diff
@@ -160,6 +162,9 @@ private Response createResponse(LoginFormsPages page) {
         if (accessCode != null) {
             uriBuilder.queryParam(OAuth2Constants.CODE, accessCode);
         }
+        if(loginHint != null) {
```
Comment (stianst) [inline services/src/main/java/org/keycloak/forms/login/freemarker/FreeMarkerLoginFormsProvider.java]: See above

File: services/src/test/java/org/keycloak/test/broker/oidc/AbstractOAuth2IdentityProviderTest.java [hunk 3]
```diff
@@ -121,10 +132,54 @@ public void getFederatedIdentity_responseUrlLine() {
 		Assert.assertEquals("458rtf", fi.getId());
 	}
 
-	private TestProvider getTested() {
+	@Test
```
Comment (stianst) [inline services/src/test/java/org/keycloak/test/broker/oidc/AbstractOAuth2IdentityProviderTest.java]: I'm not following this test. It doesn't seem to be a proper functional test which is what we need. The tests should have a identity provider configured with the pass login hint option enabled. Then go to the login with the login_hint param, click on the link and check that the username field is prefilled in the external IdP login screen (external IdP can be another KC realm on the same server). We should also check that if the pass login_hint option is disabled the username field is not prefilled.

**Review by stianst — CHANGES_REQUESTED** [review]

File: services/src/test/java/org/keycloak/test/broker/oidc/AbstractOAuth2IdentityProviderTest.java (hunk 3 above)
Comment (dmnboutin) [inline services/src/test/java/org/keycloak/test/broker/oidc/AbstractOAuth2IdentityProviderTest.java]: Hi, in this test I was just validating that "performLogin" service, which create the authorization url is working properly after the change I have done.
That's why, here I have testes it with an idp configured with activated/deactivated toggle on loginHint.

Can you explain, or show an exemple of the kind of test you are waiting for ? 
Thx in advance

Comment (stianst) [inline services/src/test/java/org/keycloak/test/broker/oidc/AbstractOAuth2IdentityProviderTest.java]: We require proper functional tests with driving the tests through the actual login pages. Take a look at https://github.com/keycloak/keycloak/blob/master/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/AbstractBrokerTest.java#L94. It uses Selenium/Webdriver to go to the actual login page, click on the IDP button, fill in the form on the IDP, etc..

Comment (dmnboutin) [inline services/src/test/java/org/keycloak/test/broker/oidc/AbstractOAuth2IdentityProviderTest.java]: Ok I will add these tests as soon as possible

Comment (dmnboutin) [inline services/src/test/java/org/keycloak/test/broker/oidc/AbstractOAuth2IdentityProviderTest.java]: Hi,
I have added functional test as requested

**Comment by dmnboutin** [comment]
Hi, Changes requested have been done.
Is it ok for you ?

**Review by stianst — APPROVED** [review]

**Comment by stianst** [comment]
Sent new PR with merge conflicts resolved, removed unnecessary changes to AbstractOAuth2IdentityProviderTest (no need to test URL as it's tested properly in KcOidcBrokerLoginHintTest). Updated KcOidcBrokerLoginHintTest to not pass in username to make sure login works without setting username.

Closing this PR and will merge https://github.com/keycloak/keycloak/pull/4421 with the above changes instead.

**Comment by stianst** [comment]
Thanks for your contribution and sorry it took so long to get it merged

**Comment by dmnboutin** [comment]
No problem, thanks for helping

**Pull request closed without merging**